src/WebJobs.Script.Grpc/Server/FunctionRpcService.cs

```diff
@@ -33,64 +31,48 @@ public FunctionRpcService(IScriptEventManager eventManager, ILogger<FunctionRpcS
 
         public override async Task EventStream(IAsyncStreamReader<StreamingMessage> requestStream, IServerStreamWriter<StreamingMessage> responseStream, ServerCallContext context)
         {
-            var cancelSource = new TaskCompletionSource<bool>();
-            IDictionary<string, IDisposable> outboundEventSubscriptions = new Dictionary<string, IDisposable>();
-
+            var cancelSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+            var cts = CancellationTokenSource.CreateLinkedTokenSource(context.CancellationToken);
+            CancellationTokenRegistration ctr = cts.Token.Register(static state => ((TaskCompletionSource<bool>)state).TrySetResult(false), cancelSource);
             try
             {
-                context.CancellationToken.Register(() => cancelSource.TrySetResult(false));
-                Func<Task<bool>> messageAvailable = async () =>
+                static Task<Task<bool>> MoveNextAsync(IAsyncStreamReader<StreamingMessage> requestStream, TaskCompletionSource<bool> cancelSource)
                 {
                     // GRPC does not accept cancellation tokens for individual reads, hence wrapper
                     var requestTask = requestStream.MoveNext(CancellationToken.None);
-                    var completed = await Task.WhenAny(cancelSource.Task, requestTask);
-                    return completed.Result;
-                };
+                    return Task.WhenAny(cancelSource.Task, requestTask);
+                }
 
-                if (await messageAvailable())
+                if (await await MoveNextAsync(requestStream, cancelSource))
```

Comment:
I don't think I've used a double await before, would love to learn more about this pattern and why we do this

Reply:
The reason there's a double `await` here is the `Task<Task<T>>` nature, so we're awaiting the async yield of the task itself, and then (outer) awaiting _that _ task.